### PR TITLE
Javadoc improvements

### LINKED ISSUE #189: [EPIC] Optimisations / Todo
**Description**
This is a list of possible optimisations. They are not worked out yet, some are self explanatory and others might require some research.

**Blocking for release**:
- [x] #196
- [x] #171 
- [x] #176 
- [x] #188 
- [x] #237
- [x] #217
- [x] #174
- [x] #202 
- [x] #262 

**Must for completing OKR**:
- [x] #231
- [x] #195
- [x] #287
- [x] #275
- [x] #273
- [x] #175
- [x] #208

**Bugfixes**:
- [x] #244 

**Cleanup**:
- [x] #227
- [x] #197
- [x] #40 
- [ ] Introduce Jib
- [ ] Add this project to Zeebe bom (check if this causes circular dependency issues)
- [ ] #133 
- [ ] #203 
- [ ] ~~Use a single gRPC server for both the engine and the control services~~
- [ ] #330

### LINKED ISSUE #196: Improve javadoc on public interfaces
**Description**

Not all interfaces are documented properly. We should go through these things and improve / add javadoc where necessary.

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

This PR adds / improves javadoc on the classes that will be commonly used by users.

## Related issues

closes #196 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to backport the fix

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually

Documentation:
* [x] Javadoc has been written
* [ ] The documentation is updated

### CODE REVIEW

File: assertions/src/main/java/io/camunda/zeebe/process/test/inspections/model/InspectedProcessInstance.java
Comment (korthout) [inline assertions/src/main/java/io/camunda/zeebe/process/test/inspections/model/InspectedProcessInstance.java:14]: 🔧 This documentation should probably belong in a commit message. It describes why we took this approach over another. I'd rather see this documentation help a user how to use it.

File: api/src/main/java/io/camunda/zeebe/process/test/api/RecordStreamSource.java
Comment (korthout) [inline api/src/main/java/io/camunda/zeebe/process/test/api/RecordStreamSource.java:8]: ❓ What does source mean in this context?

File: extension-testcontainer/src/main/java/io.camunda.zeebe.process.test.extension.testcontainer/ContainerizedEngine.java
Comment (korthout) [inline extension-testcontainer/src/main/java/io.camunda.zeebe.process.test.extension.testcontainer/ContainerizedEngine.java:49]: ❓ Is this necessary? Doesn't the javadoc get inherited automatically?

File: extension-testcontainer/src/main/java/io.camunda.zeebe.process.test.extension.testcontainer/ContainerizedEngine.java
Comment (korthout) [inline extension-testcontainer/src/main/java/io.camunda.zeebe.process.test.extension.testcontainer/ContainerizedEngine.java:78]: 🔧 Personal opinion:

>Reset the test engine. A reset stops the current engine, and replaces it with a new engine. Note that you'll need to redeploy your processes because it is an entirely new engine.

File: extension-testcontainer/src/main/java/io.camunda.zeebe.process.test.extension.testcontainer/EngineContainer.java
Comment (korthout) [inline extension-testcontainer/src/main/java/io.camunda.zeebe.process.test.extension.testcontainer/EngineContainer.java:11]: ❓ Is this something the users interact with? If so, please provide an example of how to use it.

File: extension/src/main/java/io/camunda/zeebe/process/test/extension/ZeebeProcessTestExtension.java
Comment (korthout) [inline extension/src/main/java/io/camunda/zeebe/process/test/extension/ZeebeProcessTestExtension.java:38]: 🔧 Maybe link this with an `@see` or so to documentation about the injected fields.

**Review by korthout — APPROVED** [review]
Thanks @remcowesterhoud 👍 I really like the improvements. It's a big change.

Please don't be discouraged by the number of comments I've written during this review. Most are small changes, others are merely suggestions that can be postponed. I don't think any of my comments are blocking this PR from being merged, but please have a look before merging.

File: assertions/src/main/java/io/camunda/zeebe/process/test/assertions/BpmnAssert.java
Comment (remcowesterhoud) [reply assertions/src/main/java/io/camunda/zeebe/process/test/assertions/BpmnAssert.java]: In theory it is a breaking change, but we do not expect users to use this method. That doesn't make it a good thing, but breaking changes are more than welcome before the first stable release (1.4.0). There will be some more soon 😄

File: assertions/src/main/java/io/camunda/zeebe/process/test/assertions/BpmnAssert.java
Comment (remcowesterhoud) [reply assertions/src/main/java/io/camunda/zeebe/process/test/assertions/BpmnAssert.java:24]: Preferably never, because it is taken care of by the test extensions.

In reality there is one case it could happen. If the user for some reason start multiple threads in his tests it needs to initialise a record stream for each thread., since it it is stored in a `ThreadLocal`

File: extension-testcontainer/src/main/java/io.camunda.zeebe.process.test.extension.testcontainer/EngineContainer.java
Comment (remcowesterhoud) [reply extension-testcontainer/src/main/java/io.camunda.zeebe.process.test.extension.testcontainer/EngineContainer.java:11]: It's not something for users to interact with.

File: api/src/main/java/io/camunda/zeebe/process/test/api/InMemoryEngine.java
Comment (remcowesterhoud) [reply api/src/main/java/io/camunda/zeebe/process/test/api/InMemoryEngine.java]: I just fixed this 😫 

Thanks though I'll change it 😄

**Pull request merged**
